lib/pavilion/system_plugins.py

```diff
@@ -0,0 +1,106 @@
+import collections
+from pavilion.variables import VariableSetManager
+from yapsy.IPlugin import IPlugin
+import logging
+import re
+
+LOGGER = logging.getLogger('pav.{}'.format(__name__))
+
+class PluginSystemError(RuntimeError):
+    pass
+
+_SYSTEM_PLUGINS = {}
+
+class SysVarDict( collections.UserDict ):
+
+    def __init__( self, want_deferred=True ):
+        global._SYSTEM_PLUGINS
+        super().__init__(_SYSTEM_PLUGINS)
+        self.want_deferred = want_deferred
+
+    def __getitem__( self, name ):
+        plugin = self.data[ name ]
+        return plugin.get()
+
+def add_system_plugin( system_plugin ):
+    name = system_plugin.name
+
+    if name not in _SYSTEM_PLUGINS:
+        _SYSTEM_PLUGINS[ name ] = system_plugin
+    elif priority > _SYSTEM_PLUGINS[name].priority:
+        _SYSTEM_PLUGINS[ name ] = system_plugin
+    elif priority == _SYSTEM_PLUGINS[name].priority:
+        raise PluginSystemError("Two plugins for the same system plugin have "
+                                "the same priority {}, {}."
+                                .format(system_plugin, _SYSTEM_PLUGINS[name]))
+
+def remove_system_plugin( system_plugin ):
+    name = system_plugin.name
+
+    if name in _SYSTEM_PLUGINS:
+        del _SYSTEM_PLUGINS[ name ]
+
+def get_system_plugin( name ):
+    if name not in _SYSTEM_PLUGINS:
+        raise PluginSystemError("Module not found: '{}'".format(name))
+
+    return _SYSTEM_PLUGINS[ name ]
+
+class SystemPlugin(IPlugin):
+
+    PRIO_DEFAULT = 0
+    PRIO_COMMON = 10
+    PRIO_USER = 20
+    global._SYSTEM_PLUGINS
+
+    def __init__(self, plugin_name, priority=PRIO_DEFAULT, is_deferable=False,
+                 sub_keys=[ None ]):
+        """Initialize the system plugin instance.  This should be overridden in
+        each final plugin.
+        :param str plugin_name: The name of the system plugin being wrapped.
+        :param int priority: Priority value of plugin when two plugins have
+                             the same name.
+        :param bool is_deferable: Whether the plugin is able to be deferred.
+        :param str/dict sub_keys: Key or list of keys used with this plugin.
+        """
+        super().__init__()
+
+        self.is_deferable = is_deferable
+
+        if self.NAME_VERS_RE.match(name) is None:
+            raise PluginSystemError("Invalid module name: '{}'".format(name))
+
+        self.name = plugin_name
+        self.priority = priority
+        self.sub_keys = sub_keys
+        self.values = {}
+
+        _SYSTEM_PLUGINS[ self.name ] = {}
```

Comment:
Setting this should be reserved for the activate method.

Reply:
This is especially true because we don't even know if this plugin will fully load, so you end up with modifications by plugins that fail to initialize completely.

Reply:
Also, there's a race condition if you have multiple plugins with the same name.